kotlinx-coroutines-core/common/src/flow/operators/Chunk.kt

```diff
@@ -0,0 +1,64 @@
+package kotlinx.coroutines.flow.operators
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.flow
+
+
+fun <T, R> Flow<T>.chunked(size: Int, transform: suspend (List<T>) -> R): Flow<R> = nonOverlappingWindowed(size, size, true, transform)
+
+fun <T, R> Flow<T>.windowed(size: Int, step: Int, partialWindows: Boolean, transform: suspend (List<T>) -> R): Flow<R> =
+        if (size <= step) nonOverlappingWindowed(size, step, partialWindows, transform)
+        else overlappingWindowed(size, step, partialWindows, transform)
+
+
+private fun <T, R> Flow<T>.nonOverlappingWindowed(size: Int, step: Int, partialWindows: Boolean, transform: suspend (List<T>) -> R): Flow<R> {
+
+    return flow {
+        require(size in 1..step)
+        val window = ArrayList<T>(size)
+        val toSkip = step - size
+        var skipped = toSkip
+
+        collect { value ->
+            if (toSkip == skipped) {
+                window.add(value)
+            } else skipped++
+
+            if (window.size == size) {
+                emit(transform(window.toList()))
```

Comment:
The current implementation provides a copy of the current `window`, this force a new list on each iteration.

`Sequence` and `Iterable` pass an ephemeral list to `transform`'s argument, I propose to apply the same criteria.

A not ephemeral list should be provided only if the `transformer` is missing, so `flow(...).windowed(10).toList()` is a valid statement.

Will be always possible to emulate the current behaviour with: `flow(...).windowed(...).map(...)`.

Reply:
Done